File: lib/turbofan/runtime/input_resolver.rb
require "json"

module Turbofan
  module Runtime
    module InputResolver
      def self.call(context)
        raw = deserialize(context)
        normalize_envelope(raw)
      end

      def self.deserialize(context)
        if context.array_index
          bucket = ENV.fetch("TURBOFAN_BUCKET", "turbofan-data")
          step_name = ENV.fetch("TURBOFAN_STEP_NAME")
          FanOut.read_input(
            array_index: context.array_index,
            s3_client: context.s3,
            bucket: bucket,
            execution_id: context.execution_id,
            step_name: step_name,
            chunk: context.size,
            parent_index: ENV["TURBOFAN_PARENT_INDEX"]
          )
        elsif ENV.key?("TURBOFAN_PREV_STEPS")
          fetch_parallel_outputs(context)
        elsif ENV.key?("TURBOFAN_PREV_STEP")
          fetch_previous_step_output(context)
        else
          raw = ENV.fetch("TURBOFAN_INPUT", "{}")
          parsed = JSON.parse(raw)
          parsed = Payload.deserialize(parsed, s3_client: context.s3)
          # Resolve items_s3_uri for first-step triggers (e.g., Lambda steps
          # that are the first step in a pipeline receive the trigger input
          # directly, without a chunking Lambda to resolve S3 items first).
          if parsed.is_a?(Hash) && parsed.key?("items_s3_uri")
            resolve_items_s3_uri(parsed["items_s3_uri"], context)
          else
            parsed
          end
        end
      end
      private_class_method :deserialize

      def self.fetch_previous_step_output(context)
        prev_step = ENV["TURBOFAN_PREV_STEP"]
        bucket = ENV.fetch("TURBOFAN_BUCKET", "turbofan-data")

        if ENV.key?("TURBOFAN_PREV_FAN_OUT_SIZES")
          size_names = ENV["TURBOFAN_PREV_FAN_OUT_SIZES"].split(",")
          chunks = {}
          size_names.each do |size|
            parents = JSON.parse(ENV["TURBOFAN_PREV_FAN_OUT_SIZE_#{size.upcase}"])
            parents.each do |parent|
              chunks["#{size}/parent#{parent["index"]}"] = parent["size"]
            end
          end
          FanOut.collect_outputs(
            s3_client: context.s3,
            bucket: bucket,
            execution_id: context.execution_id,
            step_name: prev_step,
            chunks: chunks
          )
        elsif ENV.key?("TURBOFAN_PREV_FAN_OUT_PARENTS")
          parents = JSON.parse(ENV["TURBOFAN_PREV_FAN_OUT_PARENTS"])
          parent_chunks = parents.each_with_object({}) { |p, h| h["parent#{p["index"]}"] = p["size"] }
          FanOut.collect_outputs(
            s3_client: context.s3,
            bucket: bucket,
            execution_id: context.execution_id,
            step_name: prev_step,
            chunks: parent_chunks
          )
        elsif ENV.key?("TURBOFAN_PREV_FAN_OUT_SIZE")
          count = ENV["TURBOFAN_PREV_FAN_OUT_SIZE"].to_i
          FanOut.collect_outputs(
            count: count,
            s3_client: context.s3,
            bucket: bucket,
            execution_id: context.execution_id,
            step_name: prev_step
          )
        else
          key = FanOut.s3_key(context.execution_id, prev_step, "output.json")
          response = Turbofan::Retryable.call { context.s3.get_object(bucket: bucket, key: key) }
          JSON.parse(response.body.read)
        end
      end
      private_class_method :fetch_previous_step_output

      def self.fetch_parallel_outputs(context)
        prev_steps = ENV["TURBOFAN_PREV_STEPS"].split(",")
        bucket = ENV.fetch("TURBOFAN_BUCKET", "turbofan-data")
        prev_steps.map do |prev_step|
          key = FanOut.s3_key(context.execution_id, prev_step, "output.json")
          response = Turbofan::Retryable.call { context.s3.get_object(bucket: bucket, key: key) }
          JSON.parse(response.body.read)
        end
      end
      private_class_method :fetch_parallel_outputs

      def self.resolve_items_s3_uri(uri, context)
        bucket = ENV.fetch("TURBOFAN_BUCKET", "turbofan-data")
        key = uri.sub("s3://#{bucket}/", "")
        response = Turbofan::Retryable.call { context.s3.get_object(bucket: bucket, key: key) }
        JSON.parse(response.body.read)
      end
      private_class_method :resolve_items_s3_uri

      def self.normalize_envelope(raw)
        if raw.is_a?(Array)
          {"inputs" => raw}
        elsif raw.is_a?(Hash) && raw.key?("inputs") && raw["inputs"].is_a?(Array)
          raw
        elsif raw.is_a?(Hash) && raw.key?("items") && raw["items"].is_a?(Array)
          {"inputs" => raw["items"]}
        else
          {"inputs" => [raw]}
        end
      end
      private_class_method :normalize_envelope
    end
  end
end

File: python/turbofan_runtime/input_resolver.py
"""Step input resolution: env var or S3 fan_out, then envelope normalize.

Mirrors lib/turbofan/runtime/input_resolver.rb. Returns a dict with at
least an "inputs" key (array). The Wrapper extracts "inputs", stashes
other envelope keys on `context.envelope`, then handles sentinel
`[None]` chunk + schema validation.

Resolution order (matches Ruby):
1. `AWS_BATCH_JOB_ARRAY_INDEX` set → fan_out path (S3 items.json)
2. `TURBOFAN_PREV_STEPS` or `TURBOFAN_PREV_STEP` → NotImplementedError
   (v2 work; v1 Python steps either drive the input themselves via
   the State Machine trigger or are fan_out children)
3. `TURBOFAN_INPUT` env → JSON-parse, hydrate any `__turbofan_s3_ref`,
   resolve `items_s3_uri` if present
"""

import json
import os

from . import fan_out
from .payload import Payload
from .retryable import Retryable


class InputResolver:
    @classmethod
    def call(cls, context):
        raw = cls._resolve(context)
        return normalize_envelope(raw)

    @classmethod
    def _resolve(cls, context):
        if context.array_index is not None:
            bucket = os.environ.get("TURBOFAN_BUCKET", "turbofan-data")
            step_name = os.environ["TURBOFAN_STEP_NAME"]
            parent_index = os.environ.get("TURBOFAN_PARENT_INDEX")
            return fan_out.read_input(
                array_index=context.array_index,
                s3_client=context.s3,
                bucket=bucket,
                execution_id=context.execution_id,
                step_name=step_name,
                chunk=context.size,
                parent_index=parent_index,
            )

        if "TURBOFAN_PREV_STEPS" in os.environ or "TURBOFAN_PREV_STEP" in os.environ:
            raise NotImplementedError(
                "Python runtime v1 does not support TURBOFAN_PREV_STEP / "
                "TURBOFAN_PREV_STEPS input resolution. Use the Ruby runtime "
                "for steps that depend on parallel or routed prev-step "
                "outputs, or wait for the v2 Python InputResolver expansion."
            )

        raw_json = os.environ.get("TURBOFAN_INPUT", "{}")
        parsed = json.loads(raw_json)
        parsed = Payload.deserialize(parsed, s3_client=context.s3)
        if isinstance(parsed, dict) and "items_s3_uri" in parsed:
            return cls._resolve_items_s3_uri(parsed["items_s3_uri"], context)
        return parsed

    @staticmethod
    def _resolve_items_s3_uri(uri, context):
        """Fetch and JSON-parse an s3:// URI from a trigger payload.

        Mirrors `input_resolver.rb#resolve_items_s3_uri`. Bucket is
        extracted from the URI itself (NOT TURBOFAN_BUCKET) so cross-
        bucket trigger payloads work correctly. (Ruby has the same
        behavior via `uri.sub("s3://#{bucket}/", "")` which only works
        when bucket matches; the urlparse approach here is strictly
        more correct.)
        """
        from urllib.parse import urlparse

        parsed = urlparse(uri)
        bucket = parsed.netloc
        key = parsed.path.lstrip("/")
        response = Retryable.call(
            lambda: context.s3.get_object(Bucket=bucket, Key=key),
            logger=context.logger,
        )
        return json.loads(response["Body"].read())


def normalize_envelope(raw):
    """Coerce arbitrary input into `{"inputs": [...]}` shape.

    Mirrors input_resolver.rb#normalize_envelope:
    - Array → {"inputs": [...]}
    - Dict with "inputs" key (Array) → returned as-is
    - Dict with "items" key (Array) → rename to "inputs", preserve siblings
    - Anything else → {"inputs": [raw]}
    """
    if isinstance(raw, list):
        return {"inputs": raw}
    if isinstance(raw, dict):
        if isinstance(raw.get("inputs"), list):
            return raw
        if isinstance(raw.get("items"), list):
            result = {k: v for k, v in raw.items() if k != "items"}
            result["inputs"] = raw["items"]
            return result
    return {"inputs": [raw]}
